### PR TITLE
[GLT-2922] Check for duplicate indices on Pool validation

### PR DESCRIPTION
**Pull request by avarsava**

### CODE REVIEW

**Comment by avarsava** [comment]
> Also, there are already pools that will become invalid because of this, which means we won't be able to edit them without removing the colliding aliquots. Is that how it should be, or should we skip validating indices if the aliquots have not changed?

This change only affects the behaviour when adding or removing library aliquots from a pool. For existing pools which are now invalid, the only valid action _on library aliquots_ is to remove the offending aliquots. Attachments, QCs, and Sequencing Orders can still be added (Sequencing Orders for invalid Pools are next to go though). It was my understanding that this change ought to be a very strict one based on our conversations, but we can discuss this further.

**Comment by avarsava** [comment]
@djcooke Repopulated PoolElements on check, limited check to new conflicts, updated documentation

File: miso-service/src/main/java/uk/ac/bbsrc/tgac/miso/service/impl/ValidationUtils.java [hunk 1]
```diff
@@ -1,33 +1,23 @@
...
 
 import java.io.IOException;
 import java.math.BigDecimal;
 import java.net.MalformedURLException;
 import java.net.URL;
-import java.util.Collection;
-import java.util.Date;
-import java.util.HashSet;
-import java.util.Map;
-import java.util.Set;
```
Comment (djcooke) [inline miso-service/src/main/java/uk/ac/bbsrc/tgac/miso/service/impl/ValidationUtils.java]: Looks like you removed a bunch of imports that are still needed?

Comment (avarsava) [reply miso-service/src/main/java/uk/ac/bbsrc/tgac/miso/service/impl/ValidationUtils.java]: I used Optimize Imports, it imports java.util.* instead

File: miso-service/src/main/java/uk/ac/bbsrc/tgac/miso/service/impl/DefaultPoolService.java [hunk 2]
```diff
@@ -266,12 +269,52 @@ public long update(Pool pool) throws IOException {
...
+  private void refreshPoolElements(Pool pool, Collection<ValidationError> errors){
+    Set<PoolElement> pes = new HashSet<>();
+    for(PoolElement oldPe : pool.getPoolContents()){
+      PoolElement newPe = new PoolElement();
+      newPe.setPool(pool);
+      newPe.setProportion(oldPe.getProportion());
+      try {
+        newPe.setPoolableElementView(poolableElementViewService.get(oldPe.getPoolableElementView().getAliquotId()));
+      } catch (IOException e) {
+        errors.add(new ValidationError("poolElements", "Failed to reconstruct PoolableElementView"));
```
[the remark this answers is not in the document]
Comment (avarsava) [reply miso-service/src/main/java/uk/ac/bbsrc/tgac/miso/service/impl/DefaultPoolService.java]: fixed

File: docs/user_manual/pools.md [hunk 3]
```diff
@@ -114,6 +114,10 @@ and select the aliquots that you wish to add to the pool. You can search aliquot
 the table. Once you have selected the appropriate aliquots, click the "Add" button in the toolbar at the top of the
 table to add the library aliquots to the pool.
 
+If you (or your administrator) have enabled Strict Index Conflict rules in 
+[miso.properties](2019-05-03-baremetal-installation-guide.md), Library Aliquots cannot be added to Pools when the
```
[the remark this answers is not in the document]
Comment (avarsava) [reply docs/user_manual/pools.md]: fixed

File: miso-web/src/main/resources/external/miso.properties [hunk 4]
```diff
@@ -90,3 +90,6 @@ miso.display.library.bulk.libraryalias:true
 miso.display.library.bulk.description:false
 miso.display.library.bulk.volume:false
 
+## Strict Index Conflict rules for adding Library Aliquots to Pools.
+## When enabled, Library Aliquots with duplicate or near-duplicate indices cannot be added to Pools.
+miso.pools.strict:false
```
[the remark this answers is not in the document]
Comment (avarsava) [reply miso-web/src/main/resources/external/miso.properties]: fixed

File: miso-service/src/main/java/uk/ac/bbsrc/tgac/miso/service/impl/ValidationUtils.java (hunk 1 above)
Comment (djcooke) [reply miso-service/src/main/java/uk/ac/bbsrc/tgac/miso/service/impl/ValidationUtils.java]: Oh, didn't see that. Please do not use that option though as the Eclipse settings in the repository do not use it, and are meant to minimize the diff between users

File: docs/user_manual/site_configuration.md [hunk 5]
```diff
@@ -190,3 +190,11 @@ easier to find that they are related because they are connected to the same ghos
 
 {% include userman-toplink.md %}
 
+{% assign sub = sub | plus: 1 %}
+{% include userman-heading.md section=page.section sub=sub title="Strict Index Checking" %}
```
Comment (djcooke) [inline docs/user_manual/site_configuration.md]: The subsection needs to be added added to the table of contents as well

Comment (avarsava) [reply docs/user_manual/site_configuration.md]: done

File: docs/user_manual/site_configuration.md [hunk 6]
```diff
@@ -190,3 +190,11 @@ easier to find that they are related because they are connected to the same ghos
...
+
+By default, any Library Aliquot can be added to a Pool. When Strict Index Checking is enabled, Library Aliquots
+cannot be added to a Pool when their addition would cause duplicate or near-duplicate indices to exist within the 
+Pool.
```
Comment (djcooke) [inline docs/user_manual/site_configuration.md]: Entity types (library aliquots, pools, etc.) and option names (strict index checking) shouldn't be capitalized in the user manual except in titles

File: miso-service/src/main/java/uk/ac/bbsrc/tgac/miso/service/impl/DefaultPoolService.java [hunk 7]
```diff
@@ -266,12 +269,48 @@ public long update(Pool pool) throws IOException {
     return savedId;
   }
 
+  private void refreshPoolElements(Pool pool, Collection<ValidationError> errors) throws IOException {
```
Comment (djcooke) [inline miso-service/src/main/java/uk/ac/bbsrc/tgac/miso/service/impl/DefaultPoolService.java]: Remove unused errors parameter

**Comment by djcooke** [comment]
Please add an entry to the release notes, and also an Upgrade Notes section that notes the new configuration

File: docs/user_manual/site_configuration.md (hunk 6 above)
Comment (avarsava) [reply docs/user_manual/site_configuration.md]: done

File: miso-service/src/main/java/uk/ac/bbsrc/tgac/miso/service/impl/DefaultPoolService.java (hunk 7 above)
Comment (avarsava) [reply miso-service/src/main/java/uk/ac/bbsrc/tgac/miso/service/impl/DefaultPoolService.java]: done

**Comment by avarsava** [comment]
> Please add an entry to the release notes, and also an Upgrade Notes section that notes the new configuration

@djcooke done

File: miso-service/src/main/java/uk/ac/bbsrc/tgac/miso/service/impl/ValidationUtils.java (hunk 1 above)
Comment (djcooke) [reply miso-service/src/main/java/uk/ac/bbsrc/tgac/miso/service/impl/ValidationUtils.java]: Bump

Comment (avarsava) [reply miso-service/src/main/java/uk/ac/bbsrc/tgac/miso/service/impl/ValidationUtils.java]: expanded the asterisks back out

**Comment by djcooke** [comment]
`uk.ac.bbsrc.tgac.miso.core.data` in `ValidationUtils` is the same

**Comment by avarsava** [comment]
> `uk.ac.bbsrc.tgac.miso.core.data` in `ValidationUtils` is the same

copied the imports from develop

**Pull request merged**